packages/cli/src/lib/snapshot/snapshotReporter.ts

```diff
@@ -57,6 +92,112 @@ export class SnapshotReporter {
     );
   }
 
+  private reportHandlers: SnapshotReporterHandlers = {
+    [SnapshotReportStatus.SUCCESS]: () => {},
+    [SnapshotReportStatus.NO_CHANGES]: () => {},
+    [SnapshotReportStatus.MISSING_VAULT_ENTRIES]: () => {},
+    [SnapshotReportStatus.ERROR]: () => {},
+  };
+
+  public setReportHandler<T extends SnapshotReportStatus>(
+    status: T,
+    handler: SnapshotReporterHandlers[T]
+  ): SnapshotReporter {
+    this.reportHandlers[status] = handler;
+    return this;
+  }
+
+  public async handleReport(): Promise<void> {
+    const reportStatuses = this.getReportStatuses();
+    for (const handler of reportStatuses.fixables) {
+      await this.callAndReset(handler);
+    }
+    if (reportStatuses.errors.length > 0) {
+      for (const handler of reportStatuses.errors) {
+        await this.callAndReset(handler);
+      }
+      return;
+    }
+    if (reportStatuses.successes.length > 0) {
+      for (const handler of reportStatuses.successes) {
+        await this.callAndReset(handler);
+      }
+    }
+  }
+
+  private async callAndReset(handler: SnapshotReportStatus) {
+    await this.reportHandlers[handler].apply(this);
+    this.reportHandlers[handler] = () => {};
+  }
+
+  private getReportStatuses() {
+    const statuses: Record<
+      'fixables' | 'successes' | 'errors',
+      SnapshotReportStatus[]
+    > = {
+      fixables: [],
+      successes: [],
+      errors: [],
+    };
+
+    this.computeMissingVaultEntries();
+    if (this.missingVaultEntriesSet.size > 0) {
+      statuses.fixables.push(SnapshotReportStatus.MISSING_VAULT_ENTRIES);
+    }
+
+    if (!this.hasChangedResources()) {
+      statuses.successes.push(SnapshotReportStatus.NO_CHANGES);
+    }
+
+    if (this.isSuccessReport()) {
+      statuses.successes.push(SnapshotReportStatus.SUCCESS);
+    }
+
+    if (this.resourceInErrorCount > 0) {
```

Comment:
Note, may need to change in the future if we have another kind of error. (sofar so good?)